### PR TITLE
Order Creation: Fix product list sync on Add Product screen

### LINKED ISSUE #5725: Order Creation: Add Product only loads first page of products
While testing the Products section in Order Creation, I noticed that the Add Product screen (`AddProductToOrder view) is no longer syncing past the first page of products.

It turns out that within the `ForEach` block that generates the product list, the equality check to trigger sync is no longer working:

https://github.com/woocommerce/woocommerce-ios/blob/a3414975044ed2a1420eacbf248c6d9c1e106015/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/ProductsSection/AddProductToOrder.swift#L27-L29

Because `ProductRowViewModel` defaults to an automatically generated `UUID` for the ID, and it uses the ID to check equality, these two elements are not equal anymore. We should probably update the Equatable conformance to check the relevant properties instead of relying on the ID.

### PR DESCRIPTION
**Pull request by rachelmcr**
Closes: #5725

_If I'm AFK when this PR is reviewed, please feel free to merge, update, or close it as needed. I can also follow up with it after my AFK._

## Description

Fixes the issue in order creation where only the first page of products was syncing in the product list. The product list now syncs more products when you reach the bottom of the list, as expected.

## Changes

1. Updates the equality (`==`) method in `ProductRowViewModel` to check all of the significant view model properties, instead of relying on the ID. Now, if two view models have the same product properties (e.g. name, imageURL), the same quantity, and the same ability to change the quantity, they are equal.
2. Updates `ProductRowViewModel` to base its default ID on the product ID by default. This avoids issues with the ID being dynamically generated, such as the `List` scrolling to the top when the data changes. The product ID will provide a unique ID for the list of products on the store, and can still be set to something unique when needed (such as on the `NewOrder` screen, where a product row can be created multiple times for the same product).

## Testing

1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab and tap the + button. Tap "Create order" in the bottom sheet.
4. Tap "Add product" and confirm when you scroll to the bottom of the product list, the next page of products is synced and loaded smoothly (without scrolling you to the top of the product list).
5. Finish adding products and create the order, and confirm everything works as expected (e.g. adding, changing quantity, removing products).

## Screenshots

https://user-images.githubusercontent.com/8658164/146971270-551f40eb-1f0f-4ea7-9a0a-1d8efb0dcd6b.mp4

## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductRowViewModel.swift [hunk 1]
```diff
@@ -174,8 +174,14 @@ private extension ProductRowViewModel {
...
+extension ProductRowViewModel: Equatable {
     static func == (lhs: ProductRowViewModel, rhs: ProductRowViewModel) -> Bool {
-        lhs.id == rhs.id
+        return lhs.productID == rhs.productID &&
+        lhs.imageURL == rhs.imageURL &&
+        lhs.name == rhs.name &&
+        lhs.stockAndPriceLabel == rhs.stockAndPriceLabel &&
+        lhs.skuLabel == rhs.skuLabel &&
+        lhs.quantity == rhs.quantity &&
+        lhs.canChangeQuantity == rhs.canChangeQuantity
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductRowViewModel.swift]: I wonder if we should group all or most of these properties on a `struct`, something like `ProductDetails`.
And then only compare that product details type, that will prevent us on forgetting to update the custom equality when updating the view model.

Comment (Ecarrion) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductRowViewModel.swift]: Shouldn't we be comparing the generated `id` for any scenario? Is it not needed on the new order screen?

**Review by Ecarrion — COMMENTED** [review]
On my testing this works but I have some problems understanding the logic that these changes are involving. 

I wonder if we should change the approach and make the view even dumber, something like.

```swift
List {
}
.onAppear {
    viewModel.bottomOfTheListReached() // This can be figured out via the scroll reader proxy
}
```

then

```swift
func bottomOfTheListReached() {
    guard morePagesTofetch() else { return }
    fetchMorePages()
}
```

That way you won't have to deal with custom equalities that could restrict the way we depend on equality as this module grows.

What do you think?

**Comment by rachelmcr** [comment]
> I wonder if we should change the approach and make the view even dumber

Yeah, that makes sense. I don't see a way to use the scroll reader proxy for this (I don't think it detects scroll position; it just scrolls to identifiable views) but we can still avoid depending on equality here.

As far as I have found, there are two ways to take an action if we have scrolled to the bottom of a list:

1. Using equality to tell if an element in the list is the last item in the data set used to populate the list (the approach I was taking here).
2. If there is a static view that always appears at the bottom of the list (and we can use `onAppear` on that view).

If we'd rather use the second approach, we can use the `ProgressView` that's at the bottom of the list. In that case we can't use our `renderedIf` view modifier to display the view, since that removes it from the view hierarchy entirely, but we could use a modifier that to just hide it when it shouldn't be shown. Something like this:

```
List {
    ForEach {
    }

    ProgressView()
        .hiddenIf(!viewModel.shouldShowScrollIndicator)
        .onAppear {
            viewModel.syncNextPage()
        }
}
```

And then we shouldn't need this view model to conform to `Equatable`, so we can drop that custom equality implementation entirely. I'll work on that and we can see what we think.

**Comment by rachelmcr** [comment]
@Ecarrion Ready for another look!

As I suggested in my last comment, in b9d7fe7 I made the `ProgressView` at the bottom of the list always part of the view hierarchy (just `hidden()` when it shouldn't be shown). That way, we can trigger sync with `onAppear` on that view. (The syncing coordinator handles everything to make sure we only sync when it's needed.) Let me know if you know of a better way to do this!

With that change, we don't need `ProductRowViewModel` to conform to `Equatable`, so I removed that entirely in dbee76f.

**Review by Ecarrion — APPROVED** [review]
Nice, this looks so much more understandable!

If you want to continue improving this to make it more readable and have fewer update points you could abstract it into a view.

```swift
private struct AlwaysThereProgressView: View {

    let showContent: Bool

    var body: some View {
        if #available(iOS 15.0, *) {
            createProgressView()
                .listRowSeparator(.hidden, edges: .bottom)
        } else {
            createProgressView()
        }
    }

    @ViewBuilder func createProgressView() -> some View {
        ProgressView()
            .frame(maxWidth: .infinity, alignment: .center)
            .listRowInsets(EdgeInsets())
            .listRowBackground(Color(.listBackground))
			.opacity(showContent ? 1 : 0) //Hidden but still in view hierarchy so `onAppear` will trigger sync when needed
    }
}
```

And then just call it like
```swift
AlwaysThereProgressView(showContent: viewModel.shouldShowScrollIndicator)
    .onAppear {
        viewModel.syncNextPage()
    }
```

**Comment by rachelmcr** [comment]
Thanks, that's a great idea! I moved it to its own view in ea6987b and I'll merge this when tests pass.

**Pull request merged**